### ISSUE TITLE
win32_gdi_font tracebacks in python 3.12

### ISSUE THREAD

**Issue by ccanepa** [pyglet/pyglet#964]
traceback while running examples\text\font_comparison.py under python 3.12, same problem with pyglet 1.5 maintenance

OS: windows 11 Pro Versión	10.0.22621 compilación 22621

pyglet 2 at
(py312_p) D:\dev\pyglet2\examples\text>git log
commit 237c56ce0855f9a52839268696254bde46172912 (HEAD -> master, origin/master, origin/HEAD)
Author: Benjamin 
Date:   Sun Oct 8 15:38:58 2023 +0900

    ComputeShaderProgram: expose MAX values directly on the class. 

```
Traceback (most recent call last):
  File "D:\dev\pyglet2\examples\text\font_comparison.py", line 128, in <module>
    window.on_text_motion(None)
  File "D:\dev\pyglet2\examples\text\font_comparison.py", line 91, in on_text_motion
    self.head = pyglet.text.Label(face, font_size=16, y=0, anchor_y='bottom')
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\dev\pyglet2\pyglet\text\__init__.py", line 444, in __init__
    super().__init__(doc, x, y, z, width, height, anchor_x, anchor_y, multiline, dpi, batch, group, rotation)
  File "D:\dev\pyglet2\pyglet\text\__init__.py", line 241, in __init__
    super().__init__(document, width, height, multiline, dpi, batch, group)
  File "D:\dev\pyglet2\pyglet\text\layout.py", line 924, in __init__
    self.document = document
    ^^^^^^^^^^^^^
  File "D:\dev\pyglet2\pyglet\text\layout.py", line 976, in document
    self._init_document()
  File "D:\dev\pyglet2\pyglet\text\layout.py", line 1448, in _init_document
    self._update()
  File "D:\dev\pyglet2\pyglet\text\layout.py", line 1362, in _update
    lines = self._get_lines()
            ^^^^^^^^^^^^^^^^^
  File "D:\dev\pyglet2\pyglet\text\layout.py", line 1339, in _get_lines
    glyphs = self._get_glyphs()
             ^^^^^^^^^^^^^^^^^^
  File "D:\dev\pyglet2\pyglet\text\layout.py", line 1496, in _get_glyphs
    glyphs.extend(font.get_glyphs(text[start:end]))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\dev\pyglet2\pyglet\font\base.py", line 350, in get_glyphs
    glyph_renderer = self.glyph_renderer_class(self)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\dev\pyglet2\pyglet\font\win32.py", line 75, in __init__
    self._create_bitmap(width, height)
  File "D:\dev\pyglet2\pyglet\font\win32.py", line 342, in _create_bitmap
    gdiplus.GdipCreateBitmapFromScan0(width, height, width * 4,
ctypes.ArgumentError: argument 5: TypeError: expected LP_c_ubyte instance instead of c_byte_Array_3520
```

with pyglet 1.5-maintenance, version 1.5.27 similar traceback while running 
examples/hello_world.py

````
Traceback (most recent call last):
  File "D:\dev\pyglet15\examples\hello_world.py", line 4, in <module>
    label = pyglet.text.Label('Hello, world!',
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "d:\venvs\py312_cocos\Lib\site-packages\pyglet\text\__init__.py", line 469, in __init__
    super(Label, self).__init__(document, x, y, width, height,
  File "d:\venvs\py312_cocos\Lib\site-packages\pyglet\text\__init__.py", line 268, in __init__
    super(DocumentLabel, self).__init__(document,
  File "d:\venvs\py312_cocos\Lib\site-packages\pyglet\text\layout.py", line 858, in __init__
    self.document = document
    ^^^^^^^^^^^^^
  File "d:\venvs\py312_cocos\Lib\site-packages\pyglet\text\layout.py", line 887, in document
    self._init_document()
  File "d:\venvs\py312_cocos\Lib\site-packages\pyglet\text\layout.py", line 1336, in _init_document
    self._update()
  File "d:\venvs\py312_cocos\Lib\site-packages\pyglet\text\layout.py", line 1247, in _update
    lines = self._get_lines()
            ^^^^^^^^^^^^^^^^^
  File "d:\venvs\py312_cocos\Lib\site-packages\pyglet\text\layout.py", line 1224, in _get_lines
    glyphs = self._get_glyphs()
             ^^^^^^^^^^^^^^^^^^
  File "d:\venvs\py312_cocos\Lib\site-packages\pyglet\text\layout.py", line 1384, in _get_glyphs
    glyphs.extend(font.get_glyphs(text[start:end]))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "d:\venvs\py312_cocos\Lib\site-packages\pyglet\font\base.py", line 401, in get_glyphs
    glyph_renderer = self.glyph_renderer_class(self)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "d:\venvs\py312_cocos\Lib\site-packages\pyglet\font\win32.py", line 100, in __init__
    self._create_bitmap(width, height)
  File "d:\venvs\py312_cocos\Lib\site-packages\pyglet\font\win32.py", line 324, in _create_bitmap
    gdiplus.GdipCreateBitmapFromScan0(width, height, width * 4,
ctypes.ArgumentError: argument 5: TypeError: expected LP_c_ubyte instance instead of c_byte_Array_3520
```

**Comment by ccanepa** [comment]
pyglet 1.5-maintenance, version 1.5.27 with python 3.11 runs fine

**Comment by ccanepa** [comment]
examples\text\font_comparison.py in master, as in OP, does not traceback with python 3.11

but some glyps look croped

**Comment by ccanepa** [comment]
I suspect ctypes  < 3.12 promoted arrays to pointer, and in 3.12 got stricter and traceback.

Could not find any reference to such change, probably will ask at some python lists.

As prof of concept I introduced a ctypes cast in the offending line (in the pyglet 2+ version) and the demo script run was as good as with python 3.11

The change:
```
(py312_p) D:\dev\pyglet2\examples\text>git diff
diff --git a/pyglet/font/win32.py b/pyglet/font/win32.py
index 67ea3c5c..3bc749db 100644
--- a/pyglet/font/win32.py
+++ b/pyglet/font/win32.py
@@ -340,7 +340,7 @@ class GDIPlusGlyphRenderer(Win32GlyphRenderer):
         self._bitmap = ctypes.c_void_p()
         self._format = PixelFormat32bppARGB
         gdiplus.GdipCreateBitmapFromScan0(width, height, width * 4,
-            self._format, self._data, ctypes.byref(self._bitmap))
+            self._format, cast(self._data, POINTER(ctypes.c_ubyte)), ctypes.byref(self._bitmap))

         self._graphics = ctypes.c_void_p()
         gdiplus.GdipGetImageGraphicsContext(self._bitmap,
```

Needs test in py38 and py311, and test the same in 1.5.maintenance.

**Comment by ccanepa** [comment]
note: it would be interesting to run examples/hello_world.py in pyglet 1.5-maintenance, with python 3.12 in linux, to see if the ctypes behavior is the same.

I don't have at hand a linux machine, if someone can report that would be great.

**Comment by pushfoo** [comment]
> but some glyphs look cropped.

> I suspect ctypes < 3.12 promoted arrays to pointer, and in 3.12 got stricter and traceback.

> Could not find any reference to such change, probably will ask at some python lists.

If the pointer suspicion doesn't end up being fulfilled, I think it's also worth looking for off-by-one errors and alpha / masking. I've had similar issues with my own font rendering experiments.

> I don't have at hand a linux machine, if someone can report that would be great.

The release tag, or specific commit? Let me know and I can give it a try after building it.

**Comment by ccanepa** [comment]
@pushfoo : thanks for the offer, but my asking was dumb: the codepath is only windows, so no testting using pyglet code.

It would be posible to make a test calling some function on glibc, but that would be much more involved.

Night!

**Comment by caffeinepills** [comment]
Thanks for looking into this. Yeah it does seem like that change to 3.12 to fix the `wintypes.BYTES` to `c_ubyte` was the culprit. 

According to the windows definitions it **_should_** be an unsigned byte (`c_ubyte`). However, this probably will break version compatibility wherever people used ctypes argtype definitions with a pointer to `wintypes.BYTE`. 

I guess the workaround for now is to just specify `c_ubyte` as the type for data and POINTERS and abandon wintypes.BYTE for compatibility sake, that way it works across all versions.

All uses of `POINTER(BYTE)` should probably be evaluated in some way, but from initial pass, it doesn't look too widespread on things we actually use. Not sure the best route, to explicitly use `c_ubyte` in its place, create a custom definition, or a ~~python version check~~.

For example:
`self._data = (ctypes.c_byte * (4 * width * height))()`
will become
`self._data = (ctypes.c_ubyte * (4 * width * height))()`

And:
`gdiplus.GdipCreateBitmapFromScan0.argtypes = [c_int, c_int, c_int, c_int, POINTER(BYTE), c_void_p]`
to
`gdiplus.GdipCreateBitmapFromScan0.argtypes = [c_int, c_int, c_int, c_int, POINTER(ctypes.c_ubyte), c_void_p]`

**Comment by caffeinepills** [comment]
Check out my pull request, and let me know if it resolves your issues. Thanks @ccanepa If so, we'll get that merged in.

**Comment by caffeinepills** [comment]
Fixed in 1.5.28
